zerver/lib/user_topics.py

```diff
@@ -95,39 +98,58 @@ def set_topic_mutes(
         recipient_id = stream.recipient_id
         assert recipient_id is not None
 
-        add_topic_mute(
+        set_user_topic_visibility_policy(
             user_profile=user_profile,
             stream_id=stream.id,
             recipient_id=recipient_id,
             topic_name=topic_name,
-            date_muted=date_muted,
+            visibility_policy=UserTopic.MUTED,
+            last_updated=date_muted,
         )
 
 
-def add_topic_mute(
+@transaction.atomic
+def set_user_topic_visibility_policy(
     user_profile: UserProfile,
     stream_id: int,
-    recipient_id: int,
     topic_name: str,
-    date_muted: Optional[datetime.datetime] = None,
+    *,
+    visibility_policy: int,
+    recipient_id: int,
+    last_updated: Optional[datetime.datetime] = None,
     ignore_duplicate: bool = False,
 ) -> None:
-    if date_muted is None:
-        date_muted = timezone_now()
-    UserTopic.objects.bulk_create(
-        [
-            UserTopic(
-                user_profile=user_profile,
-                stream_id=stream_id,
-                recipient_id=recipient_id,
-                topic_name=topic_name,
-                last_updated=date_muted,
-                visibility_policy=UserTopic.MUTED,
-            ),
-        ],
-        ignore_conflicts=ignore_duplicate,
+    if last_updated is None:
+        last_updated = timezone_now()
+
+    (row, created) = UserTopic.objects.get_or_create(
+        user_profile=user_profile,
+        stream_id=stream_id,
+        topic_name__iexact=topic_name,
+        recipient_id=recipient_id,
+        defaults={
+            "topic_name": topic_name,
+            "last_updated": last_updated,
+            "visibility_policy": visibility_policy,
+        },
     )
 
+    if not created:
+        if row.visibility_policy == visibility_policy and not ignore_duplicate:
```

Comment:
This would be better written as `if created: return` -- that early return pattern is much more readable.

I'm also not sure whether we should just return in the `row.visibility_policy == visibility_policy and ignore_duplicate` duplicate case -- there's no change, so I think changing `last_updated` is arguably wrong? Doing so as an early return might also make this more readable.

Reply:
I agree with the first point.
> if created: return -- that early return pattern is much more readable.

Regarding the second point
>  I think changing last_updated is arguably wrong? 

We are updating `last_updated` (and visibility_policy) only  when **the request is to update `visibility_policy`**

So we can simplify this to:
```python

if created: return

# Already 'muted', 'unmuted' or 'followed' case. -- duplicate request
if row.visibility_policy == visibility_policy and not ignore_duplicate:
   return JsonableError()

# existing_visibility_policy != new_visibility_policy case -- request to update visibility_policy
row.last_updated = last_updated
row.visibility_policy = visibility_policy
row.save(update_fields=["visibility_policy", "last_updated"])
```

Looks more readable?

Reply:
```python

        if created:
            return

        duplicate_request: bool = row.visibility_policy == visibility_policy

        if duplicate_request and ignore_duplicate:
            return

        if duplicate_request and not ignore_duplicate:
            visibility_policy_string: Dict[int, str] = {
                1: "muted",
                2: "unmuted",
                3: "followed",
            }
            raise JsonableError(
                _("Topic already {}").format(visibility_policy_string[visibility_policy])
            )

        row.visibility_policy = visibility_policy
        row.last_updated = last_updated
        row.save(update_fields=["visibility_policy", "last_updated"])
 ```
 Looks more readable?

Reply:
Yeah, that looks great.